### PR TITLE
reset pre-stakings in miner processing

### PR DESCRIPTION
**Pull request by ice-cronus**

### CODE REVIEW

File: tokenomics/prestaking_reset.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: ice License 1.0
+
```
Comment (ice-ares) [inline tokenomics/prestaking_reset.go]: Useless code. Delete file

File: tokenomics/contract.go [hunk 2]
```diff
@@ -202,6 +203,7 @@ type (
 	}
 	Processor interface {
 		Repository
+		IsPreStakingEndpointDisabled(ctx context.Context) error
```
Comment (ice-ares) [inline tokenomics/contract.go]: Delete this

File: tokenomics/contract.go [hunk 3]
```diff
@@ -42,6 +42,7 @@ var (
 	ErrRaceCondition                                   = errors.New("race condition")
 	ErrGlobalRankHidden                                = errors.New("global rank is hidden")
 	ErrDecreasingPreStakingAllocationOrYearsNotAllowed = errors.New("decreasing pre-staking allocation or years not allowed")
+	ErrPreStakingDisabled                              = errors.New("disabled")
```
Comment (ice-ares) [inline tokenomics/contract.go]: Delete this

File: miner/prestaking_reset.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: ice License 1.0
+
```
Comment (ice-ares) [inline miner/prestaking_reset.go]: Useless code. Delete file

File: miner/miner.go [hunk 5]
```diff
@@ -687,6 +715,11 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
 					return err
 				}
 			}
+			for _, value := range prestakingResettableUsers {
```
Comment (ice-ares) [inline miner/miner.go:715]: Useless code, delete it

File: miner/contract.go [hunk 6]
```diff
@@ -140,6 +140,12 @@ type (
 		model.DeserializedUsersKey
 	}
 
+	prestakingResettableUpdatedUser struct {
```
Comment (ice-ares) [inline miner/contract.go:144]: Delete this

File: miner/contract.go [hunk 7]
```diff
@@ -161,6 +167,7 @@ type (
 		extraBonusStartDate           *time.Time
 		extraBonusIndicesDistribution map[uint16]map[uint16]uint16
 		recalculationBalanceStartDate *time.Time
+		prestakingResetEndDate        *time.Time
```
Comment (ice-ares) [inline miner/contract.go]: Delete this

File: miner/miner.go [hunk 8]
```diff
@@ -798,3 +812,9 @@ func isAdvancedTeamEnabled(device string) bool {
 func isAdvancedTeamDisabled(device string) bool {
 	return !isAdvancedTeamEnabled(device)
 }
+
+func preStakingResetFinishedForWorker(ctx context.Context, db storage.DB, w int64, now *time.Time) (err error) {
```
Comment (ice-ares) [inline miner/miner.go]: Remove this!!!

File: miner/miner.go [hunk 9]
```diff
@@ -488,11 +492,21 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
 			if isAdvancedTeamDisabled(usr.LatestDevice) {
 				usr.ActiveT2Referrals = 0
 			}
+			prestakingReset := false
```
Comment (ice-ares) [inline miner/miner.go]: Remove this

File: miner/miner.go [hunk 10]
```diff
@@ -488,11 +492,21 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
...
+			if usr.PreStakingAllocation != 0 && usr.PreStakingBonus != 0 {
```
Comment (ice-ares) [inline miner/miner.go]: Remove this

File: miner/miner.go [hunk 11]
```diff
@@ -488,11 +492,21 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
...
+				usr.PreStakingAllocation = 0
+				usr.PreStakingBonus = 0
+				prestakingReset = true
```
Comment (ice-ares) [inline miner/miner.go]: Remove this

File: miner/miner.go [hunk 12]
```diff
@@ -488,11 +492,21 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
...
+			}
 			updatedUser, shouldGenerateHistory, IDT0Changed := mine(currentAdoption.BaseMiningRate, now, usr, t0Ref, tMinus1Ref)
 			if shouldGenerateHistory {
 				userHistoryKeys = append(userHistoryKeys, usr.Key())
 			}
 			if updatedUser != nil {
+				if prestakingReset {
```
Comment (ice-ares) [inline miner/miner.go]: remove the if

File: miner/miner.go (hunk 10 above)
Comment (ice-ares) [reply miner/miner.go]: The if

File: miner/miner.go [hunk 13]
```diff
@@ -488,11 +492,21 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
...
+			if usr.PreStakingAllocation != 0 && usr.PreStakingBonus != 0 {
+				usr.PreStakingAllocation = 0
+				usr.PreStakingBonus = 0
+				prestakingReset = true
+			}
 			updatedUser, shouldGenerateHistory, IDT0Changed := mine(currentAdoption.BaseMiningRate, now, usr, t0Ref, tMinus1Ref)
 			if shouldGenerateHistory {
 				userHistoryKeys = append(userHistoryKeys, usr.Key())
 			}
 			if updatedUser != nil {
```
Comment (ice-ares) [inline miner/miner.go:518]: U have to also set em if updatedUser is nil

File: miner/miner.go [hunk 14]
```diff
@@ -213,6 +214,9 @@ func (m *miner) mine(ctx context.Context, workerNumber int64) {
 				shouldSynchronizeBalanceFunc = m.telemetry.shouldSynchronizeBalanceFunc(uint64(workerNumber), totalBatches, iteration)
 			}
 			batchNumber = 0
+			if err := preStakingResetFinishedForWorker(ctx, m.db, workerNumber, time.Now()); err != nil {
```
Comment (ice-ares) [inline miner/miner.go]: Remove this

File: miner/miner.go (hunk 13 above)
Comment (ice-cronus) [reply miner/miner.go:518]: Is not it better with dedicated set then (like it was before)? To put there all of userResults?

Comment (ice-ares) [reply miner/miner.go:518]: Yea, it would be easier. Do that then

**Pull request merged**